### PR TITLE
Add support for `ContactPatientSheet` without appointment #1

### PR DESCRIPTION
**Pull request by Honey14**
https://app.clubhouse.io/simpledotorg/story/4298/support-contact-patient-sheet-with-no-appointment

### CODE REVIEW

File: app/src/main/java/org/simple/clinic/contactpatient/ContactPatientProfile.kt [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        parentColumn = "uuid",
+        entityColumn = "patientUuid"
+    )
+    val phoneNumbers: List<PatientPhoneNumber>,
+
+    @Relation(
+        parentColumn = "uuid",
+        entityColumn = "patientUuid"
+    )
+    val businessIds: List<BusinessId>,
```
Comment (msasikanth) [inline app/src/main/java/org/simple/clinic/contactpatient/ContactPatientProfile.kt:38]: I don't think we are using business IDs in the contact patient sheet. Can you verify that and see if we can remove this, please?

**Comment by Honey14** [comment]
Have updated the UiRenderer unit tests which are failing in checks here. Will push with the PR feedback.

File: app/src/main/java/org/simple/clinic/patient/Patient.kt [hunk 2]
```diff
@@ -364,5 +364,12 @@ data class Patient(
         AND syncStatus == 'DONE'
     """)
     abstract fun purgePatientAfterRetentionTime(now: Instant)
+
+    @Transaction
+    @Query("""
+      SELECT * FROM Patient
+      WHERE uuid = :patientUuid
+    """)
+    abstract fun contactPatientProfileImmediate(patientUuid: UUID): ContactPatientProfile
```
Comment (msasikanth) [inline app/src/main/java/org/simple/clinic/patient/Patient.kt:373]: Can we make this nullable similar to `patientProfileImmediate` function?

File: app/src/main/java/org/simple/clinic/patient/PatientRepository.kt [hunk 3]
```diff
@@ -730,6 +731,10 @@ class PatientRepository @Inject constructor(
     }
   }
 
+  fun contactPatientProfileImmediate(patientUuid: UUID): ContactPatientProfile {
```
Comment (msasikanth) [inline app/src/main/java/org/simple/clinic/patient/PatientRepository.kt:734]: Can you add a test for this, please?

File: app/src/main/java/org/simple/clinic/contactpatient/ContactPatientProfile.kt [hunk 4]
```diff
@@ -59,4 +59,13 @@ data class ContactPatientProfile(
         entityColumn = "patientUuid"
     )
     val bloodPressureMeasurement: BloodPressureMeasurement?
-) : Parcelable
+) : Parcelable {
+
+  fun withoutDeletedBusinessIds(): ContactPatientProfile {
+    return copy(businessIds = businessIds.filter { it.deletedAt == null })
+  }
```
Comment (msasikanth) [inline app/src/main/java/org/simple/clinic/contactpatient/ContactPatientProfile.kt:72]: We can remove this if we are not using business IDs in contact patient sheet.

File: app/src/sharedTest/java/org/simple/clinic/TestData.kt [hunk 5]
```diff
@@ -1320,4 +1321,67 @@ object TestData {
...
+      patientName: String = faker.name.name(),
+      patientPhoneNumber: String? = if (generatePhoneNumber) faker.phoneNumber.phoneNumber() else null,
+      businessId: BusinessId? = if (generateBusinessId) businessId(patientUuid = patientUuid) else null,
+      generateDateOfBirth: Boolean = faker.bool.bool(),
+      dateOfBirth: LocalDate? = if (generateDateOfBirth) LocalDate.parse("1980-01-01") else null,
+      age: Age? = if (!generateDateOfBirth) Age(value = kotlin.random.Random.nextInt(30..100), updatedAt = Instant.parse("2018-01-01T00:00:00Z")) else null,
+      gender: Gender = randomGender(),
+      patientDeletedReason: DeletedReason? = null,
+      patientCreatedAt: Instant = Instant.now(),
+      patientUpdatedAt: Instant = Instant.now(),
```
Comment (msasikanth) [inline app/src/sharedTest/java/org/simple/clinic/TestData.kt:1343]: Instead of `Instant.now` let's use a more deterministic values here.

File: app/src/sharedTest/java/org/simple/clinic/TestData.kt [hunk 6]
```diff
@@ -1320,4 +1321,67 @@ object TestData {
...
+      generatePhoneNumber: Boolean = faker.bool.bool(),
+      generateBusinessId: Boolean = faker.bool.bool(),
+      patientStatus: PatientStatus = PatientStatus.Active,
+      patientDeletedAt: Instant? = null,
+      patientName: String = faker.name.name(),
+      patientPhoneNumber: String? = if (generatePhoneNumber) faker.phoneNumber.phoneNumber() else null,
+      businessId: BusinessId? = if (generateBusinessId) businessId(patientUuid = patientUuid) else null,
+      generateDateOfBirth: Boolean = faker.bool.bool(),
+      dateOfBirth: LocalDate? = if (generateDateOfBirth) LocalDate.parse("1980-01-01") else null,
+      age: Age? = if (!generateDateOfBirth) Age(value = kotlin.random.Random.nextInt(30..100), updatedAt = Instant.parse("2018-01-01T00:00:00Z")) else null,
```
Comment (msasikanth) [inline app/src/sharedTest/java/org/simple/clinic/TestData.kt:1339]: We already imported `nextInt`, so we can directly use that without `Random.nextInt`

File: app/src/main/java/org/simple/clinic/contactpatient/ContactPatientModel.kt [hunk 7]
```diff
@@ -82,8 +82,8 @@ data class ContactPatientModel(
   val isPatientContactInfoLoaded: Boolean
     get() = contactPatientInfoProgressState == DONE
 
-  fun patientProfileLoaded(patientProfile: PatientProfile): ContactPatientModel {
-    return copy(patientProfile = patientProfile)
+  fun patientProfileLoaded(contactPatientProfile: ContactPatientProfile): ContactPatientModel {
```
Comment (msasikanth) [inline app/src/main/java/org/simple/clinic/contactpatient/ContactPatientModel.kt:85]: Let's rename this method name as well

File: app/src/main/java/org/simple/clinic/contactpatient/views/CallPatientView.kt [hunk 8]
```diff
@@ -147,7 +147,7 @@ class CallPatientView(
     patientAddressTextView.text = patientDetails.patientAddress
     registeredFacilityTextView.text = patientDetails.registeredFacility
     diagnosisTextView.text = diagnosis
-    lastVisitedTextView.text = dateTimeFormatter.format(patientDetails.lastVisited.toLocalDateAtZone(userClock.zone))
+    lastVisitedTextView.text = dateTimeFormatter.format(patientDetails.lastVisited?.toLocalDateAtZone(userClock.zone))
```
Comment (msasikanth) [inline app/src/main/java/org/simple/clinic/contactpatient/views/CallPatientView.kt:150]: When last visited is `null`, we should hide that view. Will that be done in a different PR?

**Review by msasikanth — CHANGES_REQUESTED** [review]
LGTM. Left few comments PTAL

File: app/src/main/java/org/simple/clinic/contactpatient/views/CallPatientView.kt (hunk 8 above)
Comment (Honey14) [reply app/src/main/java/org/simple/clinic/contactpatient/views/CallPatientView.kt:150]: Yes, doing that in a different PR.

File: app/src/main/java/org/simple/clinic/contactpatient/ContactPatientProfile.kt (hunk 4 above)
Comment (Honey14) [reply app/src/main/java/org/simple/clinic/contactpatient/ContactPatientProfile.kt:72]: We're using this in the effectHandler, ptal?

File: app/src/main/java/org/simple/clinic/patient/Patient.kt (hunk 2 above)
Comment (Honey14) [reply app/src/main/java/org/simple/clinic/patient/Patient.kt:373]: So with contactPatientProfile, I don't see a scenario where this can be null?

File: app/src/main/java/org/simple/clinic/contactpatient/ContactPatientProfile.kt (hunk 1 above)
Comment (Honey14) [reply app/src/main/java/org/simple/clinic/contactpatient/ContactPatientProfile.kt:38]: We're using this in the effectHandler, ptal?

Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/contactpatient/ContactPatientProfile.kt:38]: We are calling `withoutDeletedBusinessIds` in effect handler. But are we using business IDs from the contact patient profile somewhere? 🤔

File: app/src/main/java/org/simple/clinic/contactpatient/ContactPatientProfile.kt (hunk 4 above)
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/contactpatient/ContactPatientProfile.kt:72]: https://github.com/simpledotorg/simple-android/pull/2888#discussion_r678211732

File: app/src/main/java/org/simple/clinic/contactpatient/ContactPatientProfile.kt (hunk 1 above)
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/contactpatient/ContactPatientProfile.kt:38]: So if we are not using business IDs we can remove this and the `withoutDeletedBusinessIds` method as well.

File: app/src/main/java/org/simple/clinic/patient/Patient.kt (hunk 2 above)
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/patient/Patient.kt:373]: Hmm, good point. We expect patient to be present when calling this method. So I guess it should be fine.

File: app/src/main/java/org/simple/clinic/contactpatient/ContactPatientProfile.kt (hunk 1 above)
Comment (Honey14) [reply app/src/main/java/org/simple/clinic/contactpatient/ContactPatientProfile.kt:38]: Yes we'll filter the businessId's received from the ContactPatientProfile in the `withoutDeletedBusinessIds`, right. Hence we cannot remove them.

Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/contactpatient/ContactPatientProfile.kt:38]: Umm, okay. Why are we filtering them? Are we using it somewhere in the UI, Update or effect handler directly? At this point we are simply filtering them and not really using them right?

File: app/src/sharedTest/java/org/simple/clinic/TestData.kt (hunk 6 above)
Comment (Honey14) [reply app/src/sharedTest/java/org/simple/clinic/TestData.kt:1339]: It keeps the package by default there not sure why. can pair with you whenever possible on this.

File: app/src/androidTest/java/org/simple/clinic/patient/PatientRepositoryAndroidTest.kt [hunk 9]
```diff
@@ -3171,4 +3172,83 @@ class PatientRepositoryAndroidTest {
     val savedPrescribedDrugs = database.prescriptionDao().getAllPrescribedDrugs()
     assertThat(savedPrescribedDrugs).containsExactlyElementsIn(prescribedDrugs)
   }
+
+  @Test
+  fun when_a_patient_is_called_then_it_should_be_retrieving_contact_patient_profile_correctly() {
```
Comment (msasikanth) [inline app/src/androidTest/java/org/simple/clinic/patient/PatientRepositoryAndroidTest.kt:3177]: Can we rename this test to `loading_contact_patient_profile_should_work_correctly` or `fetching_the_contact_patient_profile_should_load_the_patient_details` instead to be consistent with other tests, please?

Comment (msasikanth) [reply app/src/androidTest/java/org/simple/clinic/patient/PatientRepositoryAndroidTest.kt:3177]: Not necessary in this PR, you can do it in the next PR.

File: app/src/sharedTest/java/org/simple/clinic/TestData.kt (hunk 6 above)
Comment (msasikanth) [reply app/src/sharedTest/java/org/simple/clinic/TestData.kt:1339]: Have you tried directly using `nextInt`? 🤔

File: CHANGELOG.md [hunk 10]
```diff
@@ -49,6 +49,7 @@
 - Bump ML Kit Barcode to v16.2.0
 - Bump Lint to v30.0.0
 - Bump AGP to v7.0.0
+- Support `ContactPatientSheet` with no appointment
```
Comment (msasikanth) [inline CHANGELOG.md:52]: If there is another PR shouldn't this be marked as in progress?

**Review by msasikanth — CHANGES_REQUESTED** [review]
LGTM. Can you just update the CHANGELOG entry to in-progress state and we can merge this in. You can resolve other 2 comments in next PR.

**Review by msasikanth — APPROVED** [review]
